keps/sig-architecture/4402-go-workspaces/README.md

```diff
@@ -0,0 +1,437 @@
+the idea of "staging" repositories - a way to eat our cake and have, too.  We
+get the benefits of a monorepo (atomic commits, fast iteration across repos)
+and then publish those to standalone repos for downstream consumption.  To make
+this work with modules, we abused GOPATH even harder and wrote even more tools.
+
+The Go project saw what we (and others) were doing and did not like it.  So
+they created [workspaces](https://go.dev/doc/tutorial/workspaces).  They
+basically created a solution that is purpose-built for us.
+
+This KEP proposes that we adopt Go workspaces and bring our tooling up to
+modern standards.  In fact, the author started this work in 2021 or 2022,
+discovering issues along the way that the Go team has worked to resolve.
+
+This is not a user-facing change.  No k8s cluster-user or cluster-admin should
+know or care that this happened.  On the surface, it seems like something that
+should not warrant a KEP, but:
+  a) KEPs are how we communicate big changes;
+  b) This is a big change;
+  c) There's some ecosystem impact to developers
+
+## Motivation
+
+Anyone who has had to maintain the multitude of `verify-*` and `update-*`
+scripts has felt this pain.  Anyone who has had to deal with how we build and
+test Kubernetes has run into the mess.  It's a constant source of complexity
+and friction for maintainers, and it leaks out into our developer ecosystem.
+
+This is the sort of mess that wise people back away from.  The author is not so
+wise.
+
+### Goals
+
+1) To get rid of our reliance on GOPATH hackery to build Kubernetes and to run
+   verifier tools and code-generators.
+2) To simplify and clarify our tooling with regards to multi-module operation.
+3) To be able to use upstream tooling, such as `gopls` across multiple modules.
+
+### Non-Goals
+
+1) To rewrite our code-generators (though some could use it).
+2) To make the build or tools faster.
+
+## Proposal
+
+This KEP proposes to add a `go.work` file (checked in) and fix all the tools
+that break because of that.  At the end, there will be no reliance on GOPATH.
+verify and update scripts can use normal Go conventions for building and
+running code.  Abominations like `run-in-gopath.sh` will be deleted.
+
+The implementation will come in several waves, which will have to be sequenced
+across k8s.io/kubernetes, k8s.io/gengo, and k8s.io/kube-openapi repos.  In
+order to not leave repos in a broken state, some PRs will need to be larger
+than we might otherwise like, but commits can tell the story.
+
+This is a proposed merge sequence (though the development PR is in a different
+order).
+
+1. Retool the core of gengo to handle workspaces.  This will be gengo/v2.  The
+   hand-rolled parsing logic will be removed and replaced with
+   `golang.org/x/tools/go/packages.Load`, which understands workspaces.
+1. Update kube-openapi to use gengo/v2.  TBD as to whether this represents
+   kube-openapi/v2 (that may be simplest).
```

Comment:
You mention later on that the API to kube-openapi wouldn't change - only the implementation. Seems to me that we don't need a separate package path in that case? Or could the new gengo result in a near-total rewrite?

Reply:
It depends on what you think of as "the API".  `kube-openapi/cmd/openapi-gen` has a `main` and so does `staging/src/k8s.io/code-generator/cmd/openapi-gen`.  In fact they are identical.  Why do we have both?  I don't know.

If we can kill off the version in `k8s.io/code-generator`, then `kube-openapi` doesn't have any "api" - it's just a tool.

Reply:
In fact, I'll see if I can do that